### DISCUSSION TITLE
How to update to Nextcloud 29 now?

### DISCUSSION THREAD

**Discussion by szaimen** [nextcloud/all-in-one#4542] [Wiki]
If you don't want to wait for AIO getting updated to NC29 for everyone in the near future, you can upgrade to Nextcloud 29 already now by following these steps:

1. First, open the aio interface and create a backup and start your containers again after the backup is successful.

2. Afterwards, run via CLI:
```
# Go into the container
sudo docker exec --user www-data -it nextcloud-aio-nextcloud bash
```

3. Now inside the container:

```
# Switch to the beta channel
php occ config:system:set updater.release.channel --value=beta

# Run the update
php updater/updater.phar --no-interaction --no-backup && php occ app:enable nextcloud-aio --force

# Switch back to the stable channel
php occ config:system:set updater.release.channel --value=stable

# Exit the container
exit
```

The above instructions should upgrade you to Nextcloud 29. You might need to follow the steps a second time to reach NC29. If something goes wrong, feel free to restore the backup.

**Comment by philwo** [comment]
I'm seeing a warning in `/settings/admin/overview` after upgrading to Nextcloud 29 saying:

"Your web server is not properly set up to resolve `.well-known` URLs, failed on: `/.well-known/caldav`."

Not sure if it causes any problems..

**Reply by afischer211** [reply]
A second warning is about not working .htaccess file...

**Reply by thefeli73** [reply]
https://docs.nextcloud.com/server/29/admin_manual/issues/general_troubleshooting.html#service-discovery

**Reply by NazrythM** [reply]
Hi.
Did you find any solution to that problem? 

I have the same problem in two instances. 
One is a fresh and clean install and the other is and upgrade, followint this discussion.

Does anyone find a solution to that warning?

I did check the .htaccess file, and everything seems fine, according to the documentation.

![imagen](https://github.com/nextcloud/all-in-one/assets/63156249/acb38451-255b-4c0c-b437-45fa101878bd)

**Reply by MeiKatz** [reply]
Not sure if this helps but the error is created [here](https://github.com/nextcloud/server/blob/master/apps%2Fsettings%2Flib%2FSetupChecks%2FWellKnownUrls.php#L93-L98). Maybe you could check which case does not work.

**Reply by dimitrij2k** [reply]
_same here... after upgrade, same error.... but everything is working with carddav/caldav on osx & sync stuff... 
btw. running on nginx.. before upgrade everything was fine..._ 

**FIXED:** https://github.com/nextcloud/all-in-one/discussions/4542#discussioncomment-9247075

**Comment by afischer211** [comment]
Inside the nextcloud container you can run php occ setupchecks. There I get additional warnings (also the two mentioned above) about PHP OpCache:
PHP opcache: The PHP OPcache module is not properly configured. OPcache is not working as it should, opcache_get_status() returns false, please check configuration.
The maximum number of OPcache keys is nearly exceeded. To assure that all scripts can be kept in the cache, it is recommended to apply "opcache.max_accelerated_files" to your PHP configuration with a value higher than "10000".

**Reply by paul-jewell** [reply]
Did you manage to resolve this? I am running into this problem as well, and even after changing the settings in /etc/php/8.2/apache2/php.ini and restarting the server, the error is still returned. For information, I am running nextcloud in a Debian VM, not docker.

Follow up: I added "opcache.enable_cli=1" to my /etc/php/8.2/apache2/conf.d/10-opcache.ini. This resolved the issue for me.

**Reply by AlphaCraft9658** [reply]
> Did you manage to resolve this? I am running into this problem as well, and even after changing the settings in /etc/php/8.2/apache2/php.ini and restarting the server, the error is still returned. For information, I am running nextcloud in a Debian VM, not docker.
> 
> Follow up: I added "opcache.enable_cli=1" to my /etc/php/8.2/apache2/conf.d/10-opcache.ini. This resolved the issue for me. 

This only applies to the native version. I don't know which is the correct .ini file for AIO.

EDIT: The OPCache is generally seemingly janky. The default config is bad. How?

**Comment by alauzon** [comment]
Nextcloud 29.0.0 is now available so no need for beta or alpha channels anymore.

**Reply by BarryS83** [reply]
Ok but even now that it has been released. How would I upgrade my Nextcloud-AIO. My master container is version v8.2.0. Nextcloud is version 28.0.4.

**Reply by alauzon** [reply]
1. First, open the aio interface and create a backup and start your containers again after the backup is successful.
2. Afterwards, run via CLI:
```
# Go into the container
sudo docker exec --user www-data -it nextcloud-aio-nextcloud bash
```
3. # Run the update
```
php updater/updater.phar --no-interaction --no-backup && php occ app:enable nextcloud-aio --force

# Exit the container
exit
```

**Reply by BarryS83** [reply]
Hi, thank for you help. I've run the posted steps. However this upgraded my Nextcloud instance from 28.0.4 to 28.0.5. But it did not upgrade it to the 29.0 version.

**Reply by ghost** [reply]
Just repeat the update-command and 29.0 should be there.

**Comment by terraboss** [comment]
`php occ setupchecks ` tells me the following after manual upgrade:

>  ✗ PHP opcache: The PHP OPcache module is not properly configured. OPcache is not working as it should, opcache_get_status() returns false, please check configuration.
> The maximum number of OPcache keys is nearly exceeded. To assure that all scripts can be kept in the cache, it is recommended to apply "opcache.max_accelerated_files" to your PHP configuration with a value higher than "10000".
> The OPcache buffer is nearly full. To assure that all scripts can be hold in cache, it is recommended to apply "opcache.memory_consumption" to your PHP configuration with a value higher than "256".
> The OPcache interned strings buffer is nearly full. To assure that repeating strings can be effectively cached, it is recommended to apply "opcache.interned_strings_buffer" to your PHP configuration with a value higher than "64"..

And it seems Nextcloud is still running under Hub 7 (28.0.5). Anyone else have this issue? I followed the steps at the beginning. (UPDATE: After running the process a second time, v29 is working, with minor errors)

<img width="1460" alt="Bildschirmfoto 2024-04-26 um 13 51 27" src="https://github.com/nextcloud/all-in-one/assets/69455/de5b140e-e944-4ebf-a1c8-89406e5b0b73">

**Reply by ghost** [reply]
Fix the "missing-indices"-error by running

`php occ db:add-missing-indices`

inside the container. Still having my issues with the other error anyway ;-) !

**Reply by terraboss** [reply]
Thank you very much, Oliver! 

🚧 Yes, sometimes you get the feeling that IT projects are a never-ending construction site ;-)

**Reply by thefeli73** [reply]
https://docs.nextcloud.com/server/29/admin_manual/issues/general_troubleshooting.html#service-discovery

**Reply by terraboss** [reply]
At the moment, this section looks like this ... Can't see the problem ... it looks good to me. 

```
<IfModule mod_rewrite.c>
  RewriteEngine on
  RewriteCond %{HTTP_USER_AGENT} DavClnt
  RewriteRule ^$ /remote.php/webdav/ [L,R=302]
  RewriteRule .* - [env=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
  RewriteRule ^\.well-known/carddav /remote.php/dav/ [R=301,L]
  RewriteRule ^\.well-known/caldav /remote.php/dav/ [R=301,L]
  RewriteRule ^remote/(.*) remote.php [QSA,L]
  RewriteRule ^(?:build|tests|config|lib|3rdparty|templates)/.* - [R=404,L]
  RewriteRule ^\.well-known/(?!acme-challenge|pki-validation) /index.php [QSA,L]
  RewriteRule ^ocm-provider/?$ index.php [QSA,L]
  RewriteRule ^(?:\.(?!well-known)|autotest|occ|issue|indie|db_|console).* - [R=404,L]
</IfModule>
```

**Comment by thefeli73** [comment]
I also have issues of warnings about `/.well-known/` problems. I am using Nginx Proxy Manager, is there anything new that needs to be added to the custom nginx config?

**Reply by thefeli73** [reply]
I fixed this issue by adding this to my NPM proxyhost config:

```
client_body_buffer_size 512k;
proxy_read_timeout 86400s;

# set max upload size and increase upload timeout:
client_max_body_size 512M;
client_body_timeout 300s;
fastcgi_buffers 64 4K;

location = /.well-known/carddav { return 301 /remote.php/dav/; }
location = /.well-known/caldav  { return 301 /remote.php/dav/; }

location /.well-known/webfinger { return 301 /index.php/.well-known/webfinger;}
location /.well-known/nodeinfo { return 301 /index.php/.well-known/nodeinfo;}

location /.well-known/acme-challenge    { try_files $uri $uri/ =404; }
location /.well-known/pki-validation    { try_files $uri $uri/ =404; }
```
The key lines are these:
```
location = /.well-known/carddav { return 301 /remote.php/dav/; }
location = /.well-known/caldav  { return 301 /remote.php/dav/; }

location /.well-known/webfinger { return 301 /index.php/.well-known/webfinger;}
location /.well-known/nodeinfo { return 301 /index.php/.well-known/nodeinfo;}
```

If you have nextcloud in a subfolder you may have to change it to this:
```
location = /.well-known/carddav { return 301 /nextcloud/remote.php/dav/; }
location = /.well-known/caldav  { return 301 /nextcloud/remote.php/dav/; }

location /.well-known/webfinger { return 301 /nextcloud/index.php/.well-known/webfinger;}
location /.well-known/nodeinfo { return 301 /nextcloud/index.php/.well-known/nodeinfo;}
```

**Reply by pintycar** [reply]
Hello @thefeli73 ! Do you have any idea what I could do to fix these errors in the case of not using any proxy? I'm simply running NextCloud AIO in docker. I never had these issues until I used the above commands to upgrade to v29. Thank you very much in advance!

**Reply by ghost** [reply]
@pintycar Just wait until the updated AIO-version is out, surely this one will be fixed then, using the official way.

**Reply by pintycar** [reply]
OK, will do, thank you @OliverPifferi  and sorry for the trouble. I thought it was officially out... I read alauzon's message from three days ago and I think I misunderstood it. Thanks again and I shall wait patiently :)

**Reply by ghost** [reply]
No problem, @pintycar / I was curious as well and reverted back to v28 after noticing these issues 😂 - feel you! 🙂

**Reply by AlphaCraft9658** [reply]
> No problem, @pintycar / I was curious as well and reverted back to v28 after noticing these issues 😂 - feel you! 🙂

I'm also reverting. Some issues may be caused by only Nextcloud itself being updated and not the companion containers.

**Reply by tim-rohrer** [reply]
Please elaborate on which file you modified.

I've updated to 29, and I've updated to the latest AIO, and I'm still getting the `well-known` URLs issue.

**Reply by NazrythM** [reply]
That modification is made it in the Nginx Proxy manager web gui.

in the specific host, in advanced options after the SSL's.

![imagen](https://github.com/nextcloud/all-in-one/assets/63156249/3443b1eb-7340-48be-b930-f6e88d0554d9)

Hope that helps.

**Reply by ShapeShifter499** [reply]
@SupayM I'm trying to switch to Nginx as my reverse proxy. I still get webfinger warnings and I've tried adding the two extra lines as below without any apparent change.

```
        location = /.well-known/webfinger { return 301 /index.php/.well-known/webfinger; }
        location = /.well-known/nodeinfo { return 301 /index.php/.well-known/nodeinfo; }
```

**Reply by ShapeShifter499** [reply]
I figured it out. For anyone that lands here, it was because I was attempting to serve custom error pages with "error_page 404 /404.html" 

The easy solution is to not use custom error pages for 404 errors. Because the code in nextcloud is looking for nextcloud itself to return what is available and it seems "webfinger" and "nodeinfo" are not present if you don't use the "social app". 

The slightly more complex workaround I found was to pass the two directly as separate proxy_pass locations with new settings that specifically exclude 'error_pages' for those two locations.  

These turn off error handling so for those two locations it gets passed directly to the nextcloud backend, no custom error pages.
```
         fastcgi_intercept_errors off;
         proxy_intercept_errors off;
         uwsgi_intercept_errors off;
```

Full example:
```
        location = /.well-known/webfinger {
         # Separate proxy_pass settings for webfinger because we want custom error_pages
         proxy_pass http://127.0.0.1:11000/index.php/.well-known/webfinger ; # Adjust to match APACHE_PORT and APACHE_IP_BINDING. See https://github.com/next>
         fastcgi_intercept_errors off;
         proxy_intercept_errors off;
         uwsgi_intercept_errors off;

         proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
         proxy_set_header X-Forwarded-Port $server_port;
         proxy_set_header X-Forwarded-Scheme $scheme;
         proxy_set_header X-Forwarded-Proto $scheme;
         proxy_set_header X-Real-IP $remote_addr;
         proxy_set_header Host $host;
         proxy_set_header Early-Data $ssl_early_data;

         # Websocket
         proxy_http_version 1.1;
         proxy_set_header Upgrade $http_upgrade;
         proxy_set_header Connection $connection_upgrade;

        }

        location = /.well-known/nodeinfo {
         # Separate proxy_pass settings for nodeinfo because we want custom error_pages
         proxy_pass http://127.0.0.1:11000/index.php/.well-known/nodeinfo ; # Adjust to match APACHE_PORT and APACHE_IP_BINDING. See https://github.com/nextc>
         fastcgi_intercept_errors off;
         proxy_intercept_errors off;
         uwsgi_intercept_errors off;

         proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
         proxy_set_header X-Forwarded-Port $server_port;
         proxy_set_header X-Forwarded-Scheme $scheme;
         proxy_set_header X-Forwarded-Proto $scheme;
         proxy_set_header X-Real-IP $remote_addr;
         proxy_set_header Host $host;
         proxy_set_header Early-Data $ssl_early_data;

         # Websocket
         proxy_http_version 1.1;
         proxy_set_header Upgrade $http_upgrade;
         proxy_set_header Connection $connection_upgrade;
        }
```

**Comment by AndrewSpec** [comment]
I've tried yesterday to force to upgrade to 29 from 28 and:

> If running with 'docker exec' try adding the option '-u 950' to the docker command (without the single quotes)
After some digging the internet I just gave up and restored it from backup.

**Reply by afischer211** [reply]
If you run docker exec with -u parameter, you should give the user "www-data". So call "docker exec -it -u www-data \<container-name\> bash". So you will go into the container as user www-data and start a bash.

**Reply by AndrewSpec** [reply]
Not the case. On the top of this discussion is a oneliner:
`sudo docker exec --user www-data -it nextcloud-aio-nextcloud bash`
And after that running the upgrade I have that message and my NC installation is broken.

**Comment by afischer211** [comment]
Does anyone has a temporarly solution for fixing the problems with htaccess, wellknown and opcache? Or should we wait for the next AIO version? Where are the files located inside running AIO container?

**Comment by tim-rohrer** [comment]
~~This seems to be working for me.~~

I too am getting the warning shown here:
![image](https://github.com/nextcloud/all-in-one/assets/7756865/00603186-b0d7-4b0f-bf7f-5440bbafda57)


I'm also curious, why doesn't updating the container image result in newer versions of NextCloud being installed?

**Comment by alemel03** [comment]
I'm running into this error:

```
00689e5f31c7:/var/www/html$ php updater/updater.phar --no-interaction --no-backup && php occ app:enable nextcloud-aio --force
Nextcloud Updater - version: v28.0.2rc3-15-gde0582d

Current version is 29.0.0.

Update to Nextcloud 29.0.0 available. (channel: "stable")
Following file will be downloaded automatically: https://download.nextcloud.com/server/releases/nextcloud-29.0.0.zip
Open changelog ↗

Updater run in non-interactive mode.

Continue update

Info: Pressing Ctrl-C will finish the currently running step and then stops the updater.

[✔] Check for expected files
[✔] Check for write permissions
[✔] Create backup
[✔] Downloading
[✔] Verify integrity
[✔] Extracting
[✔] Enable maintenance mode
[✔] Replace entry points
[✔] Delete old files
[✔] Move new files in place
[✔] Done

Update of code successful.
Updater run in non-interactive mode - will start "occ upgrade" now.

Console has to be executed with the user that owns the data directory
Current user id: 33
Owner id of the data directory: 99
Try adding 'sudo -u #99' to the beginning of the command (without the single quotes)
If running with 'docker exec' try adding the option '-u 99' to the docker command (without the single quotes)

Updater run in non-interactive mode - will disable maintenance mode now.
Console has to be executed with the user that owns the data directory
Current user id: 33
Owner id of the data directory: 99
Try adding 'sudo -u #99' to the beginning of the command (without the single quotes)
If running with 'docker exec' try adding the option '-u 99' to the docker command (without the single quotes)

Maintenance mode is disabled
```

**Comment by BarryS83** [comment]
There seem to be more issues with this 29.0 version that is breaking things on the AIO. The scheduled cron jobs als do not seem to run any longer. I am rolling back to the backup of my VM. But was wondering, will a future AIO version automatically upgrade to 29.0?

**Comment by NazrythM** [comment]
It seems to be a real problem in Nextcloud 29.

I followed all the links, and possible solutions, check every file and verifed according to the documentation.

![imagen](https://github.com/nextcloud/all-in-one/assets/63156249/fedc7eb3-a96d-4ad0-ae40-f061d2d06de7)

I have the same issue in two separated instances.

One is a balnk and clean install of NC 29, (HUB 8).

And the other one is an upgrade from NC 28.0.5, following the instructions on this thread.

Both in AIO,

If anyone finds something to fix that warning.

I wil post more updates if it produces some malfunctions (nothing until now).

Regards

Update 1:

I use this method, in the newly fresh install of the AIO with nextcloud 29.

https://github.com/nextcloud/all-in-one/discussions/4542#discussioncomment-9247075

And the warning is gone.

I will test it with the upgraded one too, and check if that is the problem.

Update 2: 

Also tested in ht eupgraded one, and it worked too, no more warnings about the well-known issue.

P.S. is good to point I'm using nginx proxy manager, in the same machine of the nextcloud AIO, both in containers and executed by a compose file as detailed in the documentation.

**Comment by gt2416** [comment]
Do you know how to fix it for users using caddy? It was working fine before the upgrade. My caddyfile looks like this  (relevant info):

# client support (e.g. os x calendar / contacts)
        redir /.well-known/carddav /remote.php/dav 301
        redir /.well-known/caldav /remote.php/dav 301
        redir /.well-known/webfinger /index.php/.well-known/webfinger 301
        redir /.well-known/nodeinfo /index.php/.well-known/nodeinfo 301

**Reply by NazrythM** [reply]
Not sure about caddy, but maybe some place where to find something similiar to NginxProxyManager.

like the image.

![imagen](https://github.com/nextcloud/all-in-one/assets/63156249/4fe5a742-5b0f-42e2-bc52-30e7ba548e26)

Hope it helps.

**Reply by Warafreak** [reply]
Try adding a trailing / to the end of these two redirect targets for WebDAV:
```
redir /.well-known/carddav /remote.php/dav/ 301
redir /.well-known/caldav /remote.php/dav/ 301
```

There seem to have been changes to the checks which did not cover all of the previously implemented tests and are thus a little too literal in some instances.

**Reply by gt2416** [reply]
> Try adding a trailing / to the end of these two redirect targets for WebDAV:
> 
> ```
> redir /.well-known/carddav /remote.php/dav/ 301
> redir /.well-known/caldav /remote.php/dav/ 301
> ```
> 
> There seem to have been changes to the checks which did not cover all of the previously implemented tests and are thus a little too literal in some instances.

Thanks for the tip. The Carddev error went away but the caldev one is still there. It still works so im ignoring these errors. Its definitely a bug rather than a setup issue so nothing else to do for now.

**Comment by szaimen** [comment]
Hi, 29 was now released for everyone running AIO with v9.0.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Reply by szaimen** [reply]
v9.0.1 stable was just released

**Comment by Delor3an91** [comment]
Hi,

1) I had perform every Nextcloud updates since with those commands at the first post of that topic but is it safe to do so or is it better to wait for containers updates and install Nextcloud updates that way?
I mean, does this CLI method for updates will not push "wrong" code for Nextcloud AIO instances instead of the packaged ones in Docker images?

2) I've got those errors after a scan that take several minutes to end and show those results (Nextcloud AIO 29.0.4):

![20240823_Nextcloud_AIO-Security_ _setup_warnings](https://github.com/user-attachments/assets/bff19a2a-1b61-4588-a6ef-b30c941c9c30)

Any help will be really appreciated.

Wish You well
